### PR TITLE
Solve_Ω at finite temperature, use it for χ0

### PR DESCRIPTION
**Pull request by antoine-levitt**
Step 1 of https://github.com/JuliaMolSim/DFTK.jl/issues/459

Depends on the Newton branch for compute_δρ. Need to profile to make sure it's not super slow, but it should be OK. Apart from that, should be good to go after the Newton PR.

### CODE REVIEW

**Comment by antoine-levitt** [comment]
The code lives in chi0.jl right now but we should probably move some stuff around (ie separate the "core" stuff about omega and K and its "users" newton and chi0) when the dust settles

File: src/postprocess/chi0.jl [hunk 1]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
     δψnk
 end
 
-"""
-Returns the change in density `δρ` for a given `δV`. Drop all non-diagonal terms with
-(f(εn)-f(εm))/(εn-εm) factor less than `droptol`. If `sternheimer_contribution`
-is false, only compute excitations inside the provided orbitals.
+# Returns δψ, δocc, δεF corresponding to a change in *total* Hamiltonian dH
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:135]: Don't you use `occupation` and thus `δoccupation` ?

File: src/postprocess/chi0.jl [hunk 2]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
-                          temperature=ham.basis.model.temperature,
-                          kwargs_sternheimer=(cgtol=1e-6, verbose=false))
+# Then <ψm|δP|ψn> = (fm-fn)/(εm-εn) <ψm|δH|ψn>
+# Except for the diagonal which is
+# <ψn|δP|ψn> = (fn'-δεF) <ψn|δH|ψn>
+
+# We want to represent this with a δψ, δf, we do *not* impose that
+# δψ is orthogonal at finite temp.
+# We get
+# δP = ∑_k fk |δψk><ψk| + |ψk><δψk| + δ fk |ψk><ψk|
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:147]: Parenthesis ? `δP = ∑_k fk (|δψk><ψk| + |ψk><δψk|) + δ fk |ψk><ψk|`

File: src/postprocess/chi0.jl [hunk 3]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+# This formula is very nice
+# - It gives a vanishing contribution fn αmn for empty states (note that α itself blows up, but it's compensated by fn)
+# - In the case where fn=1/0 or fm=0 we recover the same formulas as the ones with orthogonality
+# - It does not blow up for degenerate states
+function get_αmn(fm, fn, ratio)
+    ratio == 0 && return ratio
+    ratio * fn / (fn^2 + fm^2)
+end
+
+@views function solve_Ω(ham, ψ, occ, εF, eigenvalues, δHψ; kwargs_sternheimer=(cgtol=1e-6, verbose=false))
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:174]: I remember you telling me that you use `occupation` and not `occ` :stuck_out_tongue:

File: src/postprocess/chi0.jl [hunk 4]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+                δocc[ik][n] = (δεnk-δεF) * fpnk
+            end
+        end
+    end
+    # δεF ensures charge neutrality
+    @assert abs(mpi_sum(sum(sum, δocc), basis.comm_kpts)) < sqrt(eps(T))
+
+    # compute δψnk band per band
+    δψ = zero.(ψ)
+    for ik = 1:Nk
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:221]: `enumerate` or `map` ?

File: src/postprocess/chi0.jl [hunk 5]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
-(see `n_ep_extra` parameter). These should be discarded before using this function.
-"""
-@views @timing function apply_χ0(ham, ψ, εF, eigenvalues, δV;
-                          droptol=0,
-                          sternheimer_contribution=true,
-                          temperature=ham.basis.model.temperature,
-                          kwargs_sternheimer=(cgtol=1e-6, verbose=false))
+# Then <ψm|δP|ψn> = (fm-fn)/(εm-εn) <ψm|δH|ψn>
+# Except for the diagonal which is
+# <ψn|δP|ψn> = (fn'-δεF) <ψn|δH|ψn>
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:142]: I don't get what are `f'n` and `fn'` ?

File: src/postprocess/chi0.jl [hunk 6]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+    @assert abs(mpi_sum(sum(sum, δocc), basis.comm_kpts)) < sqrt(eps(T))
+
+    # compute δψnk band per band
+    δψ = zero.(ψ)
+    for ik = 1:Nk
+        ψk = ψ[ik]
+        δψk = δψ[ik]
+
+        εk = eigenvalues[ik]
+        for n = 1:length(eigenvalues[ik])
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:226]: `εk` ?

File: src/postprocess/chi0.jl [hunk 7]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+            fnk = filled_occ * Smearing.occupation(model.smearing, (εk[n]-εF) / temperature)
+
+            # explicit contributions
+            for m = 1:size(ψk, 2)
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:230]: Why don't use `εk` again ?

File: src/postprocess/chi0.jl [hunk 8]
```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+                αmn = get_αmn(fmk, fnk, ratio)
+                # αnm = get_αmn(fnk, fmk, ratio)
+                # @assert fnk * αmn + fmk * αnm ≈ ratio
+                δψk[:, n] .+= ψk[:, m] .* αmn .* dot(ψk[:, m], δHψ[ik][:, n])
+            end
+
+            # Sternheimer contribution
+            # TODO unoccupied orbitals may have a very bad sternheimer solve, be careful about this
+            δψk[:, n] .-= sternheimer_solver(ham.blocks[ik], ψk, ψk[:, n], εk[n], δHψ[ik][:, n];
+                                                 kwargs_sternheimer...)
```
Comment (gkemlin) [inline src/postprocess/chi0.jl:244]: alignment :stuck_out_tongue:

File: src/postprocess/chi0.jl (hunk 4 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:221]: not really because it does two things: increment D and assign docc

File: src/postprocess/chi0.jl (hunk 6 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:226]: not really simpler to do this

File: src/postprocess/chi0.jl [hunk 9]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
     δψnk
 end
 
-"""
-Returns the change in density `δρ` for a given `δV`. Drop all non-diagonal terms with
-(f(εn)-f(εm))/(εn-εm) factor less than `droptol`. If `sternheimer_contribution`
-is false, only compute excitations inside the provided orbitals.
+# Returns δψ, δocc, δεF corresponding to a change in *total* Hamiltonian dH
+# P = f_fd(H-εF), tr(P) = N
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:136]: What is `f_fd`?

Comment (mfherbst) [reply src/postprocess/chi0.jl:136]: Maybe write that P is the density matrix to make this more relatable.

File: src/postprocess/chi0.jl [hunk 10]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+# Let <ψm|δψn> = αmn <δψm|δH|ψn>. αmn has to satisfy
+# fn αmn + fm αnm = ratio = (fn-fm)/(εn-εm)
+# The usual way is to impose orthogonality (=> αmn=-αnm),
+# but this means that αmn = 1/(εm-εn), which is unstable
+# Instead, we minimize αmn^2 + αnm^2 under the linear constraint above, which leads to
+# fn αmn = ratio * fn^2 / (fn^2 + fm^2)
+
+# This formula is very nice
+# - It gives a vanishing contribution fn αmn for empty states (note that α itself blows up, but it's compensated by fn)
+# - In the case where fn=1/0 or fm=0 we recover the same formulas as the ones with orthogonality
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:168]: Lines too long.

File: src/postprocess/chi0.jl [hunk 11]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+# - It does not blow up for degenerate states
+function get_αmn(fm, fn, ratio)
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:170]: We decided to drop the `get_` prefixes and go with `compute_` instead. Also `α` is quite overloaded. Maybe use a different symbol.

File: src/postprocess/chi0.jl [hunk 12]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+# <ψn|δP|ψn> = (fn'-δεF) <ψn|δH|ψn>
+
+# We want to represent this with a δψ, δf, we do *not* impose that
+# δψ is orthogonal at finite temperature.
+# We get
+# δP = ∑_k fk (|δψk><ψk| + |ψk><δψk|) + δfk |ψk><ψk|
+# Identifying with <ψm|δP|ψn> we get for the diagonal terms
+# δfn = fn'<ψn|δH-δεF|ψn>
+# and for the off-diagonal
+# (fm-fn)/(εm-εn) <ψm|δH|ψn> = fm <δψm|ψn> + fn <ψm|δψn>
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:152]: I think I would add another paragraph break here (now you concern the explict terms with sternheimer and without)

File: src/postprocess/chi0.jl [hunk 13]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+# for the already-computed states, we add an explicit contribution
+# for the empty states, we solve a Sternheimer equation
+# (H-εn) δψn = - P_{ψ^⟂} δH ψn
+
+# The explicit term needs a careful consideration of stability.
+# Let <ψm|δψn> = αmn <δψm|δH|ψn>. αmn has to satisfy
+# fn αmn + fm αnm = ratio = (fn-fm)/(εn-εm)
+# The usual way is to impose orthogonality (=> αmn=-αnm),
+# but this means that αmn = 1/(εm-εn), which is unstable
+# Instead, we minimize αmn^2 + αnm^2 under the linear constraint above, which leads to
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:163]: above refers where?

File: src/postprocess/chi0.jl [hunk 14]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+    ratio * fn / (fn^2 + fm^2)
+end
+
+@views function solve_Ω(ham, ψ, occ, εF, eigenvalues, δHψ; kwargs_sternheimer=(cgtol=1e-6, verbose=false))
     basis  = ham.basis
+    model = basis.model
+    temperature = model.temperature
+    filled_occ = filled_occupation(model)
     T      = eltype(basis)
-    @assert basis.model.spin_polarization in (:none, :spinless, :collinear)
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:154]: I think you still assume that, don't you?

File: src/postprocess/chi0.jl [hunk 15]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+    ratio == 0 && return ratio
+    ratio * fn / (fn^2 + fm^2)
+end
+
+@views function solve_Ω(ham, ψ, occ, εF, eigenvalues, δHψ; kwargs_sternheimer=(cgtol=1e-6, verbose=false))
     basis  = ham.basis
+    model = basis.model
+    temperature = model.temperature
+    filled_occ = filled_occupation(model)
     T      = eltype(basis)
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:180]: `=` sign formatting

File: src/postprocess/chi0.jl [hunk 16]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+                δεnk = real(dot(ψ[ik][:, n], δHψ[ik][:, n]))
+                fpnk = (filled_occ 
+                        * Smearing.occupation_derivative(model.smearing, enred)
+                        / temperature)
+                δocc[ik][n] = δεnk * fpnk
+                D += fpnk * basis.kweights[ik]
+            end
+        end
+        D = mpi_sum(D, basis.comm_kpts) # equal to minus the total DOS
+        dos = sum(compute_dos(εF, basis, eigenvalues))
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:201]: Why compute twice?

File: src/postprocess/chi0.jl [hunk 17]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+    filled_occ = filled_occupation(model)
     T      = eltype(basis)
-    @assert basis.model.spin_polarization in (:none, :spinless, :collinear)
+    Nk = length(basis.kpoints)
 
-    n_spin = basis.model.n_spin_components
-    @assert 1 ≤ n_spin ≤ 2
+    # First compute δocc, δεF
+    δεF = zero(T)
+    δocc = [zero(occ[ik]) for ik = 1:Nk]
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:185]: = `fn' * δεF`

File: src/postprocess/chi0.jl [hunk 18]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+                enred = (εnk - εF) / temperature
+                δεnk = real(dot(ψ[ik][:, n], δHψ[ik][:, n]))
+                fpnk = (filled_occ 
+                        * Smearing.occupation_derivative(model.smearing, enred)
+                        / temperature)
+                δocc[ik][n] = δεnk * fpnk
+                D += fpnk * basis.kweights[ik]
+            end
+        end
+        D = mpi_sum(D, basis.comm_kpts) # equal to minus the total DOS
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:200]: Two spaces before `#`

File: src/postprocess/chi0.jl [hunk 19]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+        ψk = ψ[ik]
+        δψk = δψ[ik]
+
+        εk = eigenvalues[ik]
+        for n = 1:length(εk)
+            fnk = filled_occ * Smearing.occupation(model.smearing, (εk[n]-εF) / temperature)
+
+            # explicit contributions
+            for m = 1:length(εk)
+                n == m && continue # contribution already taken care of by δocc
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:232]: Two spaces before `#`

File: src/postprocess/chi0.jl [hunk 20]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+                fmk = filled_occ * Smearing.occupation(model.smearing, (εk[m]-εF) / temperature)
+                ddiff = Smearing.occupation_divided_difference
+                ratio = filled_occ * ddiff(model.smearing, εk[m], εk[n], εF, temperature)
+                αmn = get_αmn(fmk, fnk, ratio)
+                # αnm = get_αmn(fnk, fmk, ratio)
+                # @assert fnk * αmn + fmk * αnm ≈ ratio
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:238]: Remove?

File: src/postprocess/chi0.jl [hunk 21]
```diff
@@ -170,88 +272,9 @@ returns `3` extra bands, which are not converged by the eigensolver
...
-        rhs = r_to_G(basis, hamk.kpoint, .- δV .* ψnk_real)
-        norm(rhs) < 100eps(T) && return δρk
-        δψnk = sternheimer_solver(hamk, ψk, ψnk, εk[n], rhs; kwargs_sternheimer...)
-        δψnk_real = G_to_r(basis, hamk.kpoint, δψnk)
-        δρk .+= 2 .* fnk .* real(conj(ψnk_real) .* δψnk_real)
-    end
-
-    δρk
+    δHψ = [DFTK.RealSpaceMultiplication(basis, kpt, @views δV[:, :, :, kpt.spin]) * ψ[ik]
+       for (ik, kpt) in enumerate(basis.kpoints)]
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:276]: Indention

File: src/postprocess/chi0.jl [hunk 22]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+            # TODO unoccupied orbitals may have a very bad sternheimer solve, be careful about this
+            δψk[:, n] .-= sternheimer_solver(ham.blocks[ik], ψk, ψk[:, n], εk[n], δHψ[ik][:, n];
+                                             kwargs_sternheimer...)
+        end
+    end
+    δψ, δocc, δεF
+end
+
+"""
+Get the density variation δρ corresponding to a total potential variation δV.
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:252]: You don't want to keep the comment:
```
Note: This function assumes that all bands contained in `ψ` and `eigenvalues` are
sufficiently converged. By default the `self_consistent_field` routine of `DFTK`
returns `3` extra bands, which are not converged by the eigensolver
(see `n_ep_extra` parameter). These should be discarded before using this function.
```

File: src/postprocess/chi0.jl [hunk 23]
```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
...
+# This formula is very nice
+# - It gives a vanishing contribution fn αmn for empty states (note that α itself blows up, but it's compensated by fn)
+# - In the case where fn=1/0 or fm=0 we recover the same formulas as the ones with orthogonality
+# - It does not blow up for degenerate states
+function get_αmn(fm, fn, ratio)
+    ratio == 0 && return ratio
+    ratio * fn / (fn^2 + fm^2)
+end
+
+@views function solve_Ω(ham, ψ, occ, εF, eigenvalues, δHψ; kwargs_sternheimer=(cgtol=1e-6, verbose=false))
```
Comment (mfherbst) [inline src/postprocess/chi0.jl:175]: Should we just pass the `kwargs` to sternheimer? Now that droptol etc are gone?

File: src/postprocess/chi0.jl (hunk 11 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:170]: this is a very specific concept so it's fine to have alpha, no? I don't really have a better name...

File: src/postprocess/chi0.jl (hunk 14 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:154]: well everything right now assumes that anyway, I don't think it makes sense to have it everywhere

File: src/postprocess/chi0.jl (hunk 15 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:180]: ?

File: src/postprocess/chi0.jl (hunk 20 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:238]: I think it's nice to understand things

File: src/postprocess/chi0.jl (hunk 22 above)
Comment (antoine-levitt) [reply src/postprocess/chi0.jl:252]: good point

**Comment by antoine-levitt** [comment]
I'm not very happy with the names, and the fact that Omega^-1 = chi04P^-1 is super confusing, but let's keep it that way right now and I'll try to clean it up later when we have all the functionality

File: src/postprocess/chi0.jl (hunk 14 above)
Comment (mfherbst) [reply src/postprocess/chi0.jl:154]: I do. It helps when you are removing the assumption later

**Pull request merged**